src/components/LIcon.vue

```diff
@@ -0,0 +1,121 @@
+<template>
+    <div>
+        <slot></slot>
+    </div>
+</template>
+
+<script>
+import findRealParent from '../utils/findRealParent.js';
+import { optionsMerger } from '../utils/optionsUtils.js';
+import Options from '../mixins/Options.js';
+
+export default {
+  name: 'LIcon',
+  mixins: [Options],
+  props: {
+    iconUrl: String,
+    iconRetinaUrl: String,
+    iconSize: [Object, Array],
+    iconAnchor: [Object, Array],
+    popupAnchor: [Object, Array],
+    tooltipAnchor: [Object, Array],
+    shadowUrl: String,
+    shadowRetinaUrl: String,
+    shadowSize: [Object, Array],
+    shadowAnchor: [Object, Array],
+    bgPos: [Object, Array],
+    className: String
+  },
+
+  data() {
+    return {
+      observer: null
+    }
+  },
+
+  mounted () {
+    const options = optionsMerger({
+      iconUrl: this.iconUrl,
+      iconRetinaUrl: this.iconRetinaUrl,
+      iconSize: this.iconSize,
+      iconAnchor: this.iconAnchor,
+      popupAnchor: this.popupAnchor,
+      tooltipAnchor: this.tooltipAnchor,
+      shadowUrl: this.shadowUrl,
+      shadowRetinaUrl: this.shadowRetinaUrl,
+      shadowSize: this.shadowSize,
+      shadowAnchor: this.shadowAnchor,
+      bgPos: this.bgPos,
+      className: this.className
+    }, this);
+
+    this.watchProps(options);
+    this.watchSlot(options);
+
+    this.createIcon(options);
+  },
+
+  beforeDestroy () {
+    if (this.parentContainer.mapObject) {
+      this.parentContainer.mapObject.setIcon(null);
+    }
+
+    this.observer.disconnect();
+  },
+
+  render() {
+    return null;
+  },
+
+  methods: {
+    createIcon(options) {
+      if (this.iconObject) {
+        L.DomEvent.off(this.iconObject, this.$listeners);
+      }
+
+      options.html = this.$el.innerHTML;
+
+      if (options.html) {
+        this.iconObject = L.divIcon(options);
+      } else {
+        this.iconObject = L.icon(options);
+      }
+
+      L.DomEvent.on(this.iconObject, this.$listeners);
+
+      this.parentContainer = findRealParent(this.$parent);
+      this.parentContainer.mapObject.setIcon(this.iconObject);
+    },
+
+    watchProps(options) {
+      const props = this.$options.props;
+      const keys = Object.keys(props);
+
+      for (let i = 0; i < keys.length; i++) {
+        this.$watch(keys[i], (newVal, oldVal) => {
+          //TODO: why is watch triggered on foreign props when changing icon size in my example?
+          if (JSON.stringify(newVal) === JSON.stringify(oldVal)) {
```

Comment:
I'm relatively new to Vue, so if someone could help me out how to avoid this line, I would be very thankful!

Problem is, when sliding the slider in the example, the watcher is also triggered for the fourth icon elements properties, but it should only be called for the `iconSize` and `iconAnchor` properties of the third icon.

Reply:
Vue pass any prop written in the component template to the $props, even if not defined in the props definition.
Also you don't need this method, for an inspiration please take a look at how the function in `utils/propsBinder.js` works and how is used in any other component.

The gist of it is that you can write a function in the methods called: setPROPNAME and that function will be called only when your props change.

Reply:
Thanks for your reply and your awesome code review! 

I updated the code (marked all props as custom so the component functions will get called, because Icon doesn't have setters on an existing Icon), but when you check it out and test the example, you will notice that it still doesn't behave correct.

When changing the slider, the first icon will get recreated twice (because two props are affected by the slider), AND the second icon is also recreated (because of the iconAnchor prop). This prop shouldn't change. 
Also when you change the text, again the first icon is recreated twice (because of the iconSize and iconAnchor props) and the second one also twice (first correctly because of the html change but then also because of the iconAnchor prop).
```
------ Page load:
1541353711324 - recreate icon 1 (initial)
1541353711327 - recreate icon 2 (initial)
------ Changing slider:
1541353713991 - recreate icon 1 (iconSize)
1541353713992 - recreate icon 1 (iconAnchor)
1541353713994 - recreate icon 2 (iconAnchor)
------ Changing text:
1541353720983 - recreate icon 1 (iconSize)
1541353720985 - recreate icon 1 (iconAnchor)
1541353720986 - recreate icon 2 (iconAnchor)
1541353720988 - recreate icon 2 (html slot)
```

So two problems:
1. When two or more props on the same component are change simultanuosly, is it possible to trigger a recreation only once?
2. Why are all props detected as changed, when only one of them changes?

I hope you can help me out here too! :)

Reply:
@byWulf this could be an issue with the fact that the icon component gets recreated by his parent ( marker) you can check this by verifying that the mounted / updated lifecycle hooks get called.

there is a lot of performance issues with icon and markers but I have the feeling that you can get it right ( and I will help you to do so ) 

Really appreciate all your efforts and attitude to this!
If you want to discuss this more faster ping me on Discord, in VueLand you can find me by the name of DonNico

Reply:
@lordfuoco I spent the last night on analysing and tweaking the performance.

Problem 1 was in the example, that I didn't bind computed values to the component, but directly binded arrays (`<l-icon :icon-anchor="[0, 50]" />`). This led to vue thinking the value has changed on every check. `<l-icon :icon-anchor="iconAnchor" />` + `computed: { iconAnchor() { return [0, 50]; }}``  solved this.

Problem 2 with the "if multiple props were affected by a single change, the icon gets recreated also multiple times" was solved by scheduling a recreation on `$nextTick` and remember in a data variable, if recreation in this tick already happend, if multiple recreations were scheduled.

Also I added another tweak: If only the html content changed, I just swap the HTML in the DOM and don't recreate the icon itself.

In my real world project (880 markers, where only few of them have individual divIcons) it was recommended to put the default icons in data variables and bind them to the `icon` prop of the marker and add an `v-if` to the `l-icon` component. Works like a charme now :)

Looking forward for your feedback/code review on the current status, so we could release it soon :)